database/user.go

```diff
@@ -886,22 +886,47 @@ func (u User) HasKey(pk PubKey) bool {
 
 // monthStart returns the start of the user's subscription month.
 // Users get their bandwidth quota reset at the start of the month.
+//
+// This function follows the behaviour of Stripe:
+// If a month doesn’t have the anchor day, the subscription will be billed on
+// the last day of the month. For example, a subscription starting on 31 January
+// bills on 28 February (or 29 February in a leap year), then 31 March, 30
+// April, and so on.
+//
+// See: https://stripe.com/docs/billing/subscriptions/billing-cycle
+//
+// NOTE: This function ignores the time (hour and minutes) of the sub expiration
+// - all quotas reset at midnight UTC.
 func monthStart(subscribedUntil time.Time) time.Time {
-	now := time.Now().UTC()
-	// Check how many days are left until the end of the user's subscription
-	// month. Then calculate when the last subscription month started. We don't
-	// care if the user is no longer subscribed and their sub expired 3 months
-	// ago, all we care about here is the day of the month on which that
-	// happened because that is the day from which we count their statistics for
-	// the month. If they were never subscribed we use Jan 1st 1970 for
-	// SubscribedUntil.
-	daysDelta := subscribedUntil.Day() - now.Day()
-	monthsDelta := 0
-	if daysDelta > 0 {
-		// The end of sub day is after the current date, so the start of month
-		// is in the previous month.
-		monthsDelta = -1
-	}
-	d := now.AddDate(0, monthsDelta, daysDelta)
-	return time.Date(d.Year(), d.Month(), d.Day(), 0, 0, 0, 0, time.UTC)
+	return monthStartWithTime(subscribedUntil, time.Now().UTC())
+}
+
+// monthStartWithTime returns the start of the user's subscription month in
+// relation to the given `now` value. This function exists only for testing
+// purposes and implements the desired behaviour of monthStart.
+func monthStartWithTime(subscribedUntil time.Time, current time.Time) time.Time {
+	// Normalize the day of month. Subs ending on 31st should end on the last
+	// day of the month when the month doesn't have 31 days.
+	dayOfMonth := normalizeDayOfMonth(current.Month(), subscribedUntil.Day(), current)
+	// If we're past the reset day this month, use the current day of the month.
+	if current.Day() >= dayOfMonth {
+		return time.Date(current.Year(), current.Month(), dayOfMonth, 0, 0, 0, 0, time.UTC)
+	}
+	// If we haven't reached the reset day this month, use last month's day.
+	dayOfMonth = normalizeDayOfMonth(current.Month()-1, subscribedUntil.Day(), current)
+	return time.Date(current.Year(), current.Month()-1, dayOfMonth, 0, 0, 0, 0, time.UTC)
+}
+
+// normalizeDayOfMonth checks whether the current month has the given day and if
+// it doesn't, it returns the last day the current month has.
+//
+// Example:
+// In February normalizeDayOfMonth(31) will return 28 or 29.
+func normalizeDayOfMonth(month time.Month, day int, current time.Time) int {
```

Comment:
f/u 

This is a bit confusing. It says "whether the current month has the given day" but there is a `month` and a `current` input. So I'd either split it up into `currentMonth` and `currentYear` inputs or have `current` and then inside the function pull out the month.

Reply:
Right, that is confusing. I'll F/U.